# Tugas Kelompok Modul 1
**Nama Anggota Kelompok 8:**
1. Arafi Ramadhan Maulana - 122450004
2. Dwi Ratna Anggraeni - 122450008
3. Raid Muhammad Naufal - 122450027
4. Rayan Koemi Karuby - 122450038
5. M. Deriansyah Okutra - 1224500  



## 1. Buatlah fungsi komposisi dengan nilai f(x) = 4^x, g(x) = x^x, dan h(x) = 3 + x dengan pemrograman prosedural tiga fungsi dan pemrograman fungsional dengan lambda.



### Pemrograman Prosedural

In [39]:
import time

def f(x):
  return 4 ** x

def g(x):
  return x ** x

def h(x):
  return 3 + x

start_time = time.time()

x = 1
fungsi_komposisi = h(g(f(x)))

end_time = time.time()

waktu = end_time - start_time

print(f"Hasil dari fungsi h(g(f({x}))) = {fungsi_komposisi}")
print(f"Waktu : {waktu:.10f} detik")

Hasil dari fungsi h(g(f(1))) = 259
Waktu : 0.0000894070 detik


### Pemrograman Fungsional (Lambda)

In [35]:
import time

f = lambda x: 4 ** x
g = lambda x: x ** x
h = lambda x: 3 + x

start_time = time.time()

x = 1
fungsi_komposisi = h(g(f(x)))

end_time = time.time()

waktu = end_time - start_time

print(f"Hasil dari fungsi h(g(f({x}))) = {fungsi_komposisi}")
print(f"Waktu : {waktu:.10f} detik")

Hasil dari fungsi h(g(f(1))) = 259
Waktu : 0.0000743866 detik


## 2. Buatlah pemrograman berbasis objek pada soal 1

In [33]:
class Fungsi:
  def __init__(self, persamaan):
    self.persamaan = persamaan

  def __call__(self, x):
    return eval(self.persamaan, {"x": x})

f = Fungsi("4 ** x")
g = Fungsi("x ** x")
h = Fungsi("3 + x")

start_time = time.time()

x = 1
fungsi_komposisi = h(g(f(x)))

end_time = time.time()

waktu = end_time - start_time

print(f"Hasil dari fungsi h(g(f({x}))) = {fungsi_komposisi}")
print(f"Waktu : {waktu:.10f} detik")

Hasil dari fungsi h(g(f(1))) = 259
Waktu : 0.0001111031 detik


## 3. Berikan analisis komparasi soal 1 dan 2

Dengan menggunakan library time maka dapat diketahui perbandingan antara pemrograman prosedural, pemrograman fungsional, dan pemrogarman berbasis objek memiliki selisih waktu yang tidak berbeda jauh dan selalu berubah-ubah (tidak tepat) dengan yang tercepat adalah pemrograman fungsional (0.0000743866 detik) lalu pemrograman prosedural (0.0000894070 detik) dan pemrograman berbasis objek (0.0001111031 detik).

## 4. Buatlah satu penyelesaian Luas Segitiga dengan menggunakan pemrograman berorientasi objek dan pemrograman berbasis fungsi. Uji dengan membandingkan mana yang tercepat komputasinya.

### Pemrograman Berorientasi Objek

In [45]:
import time

class Segitiga:
  def __init__(self, alas, tinggi):
    self.alas = alas
    self.tinggi = tinggi

  def luas(self):
    return 0.5 * self.alas * self.tinggi

start_time = time.time()

alas = 10
tinggi = 16
segitiga = Segitiga(alas, tinggi)
luas_segitiga = segitiga.luas()

end_time = time.time()

waktu = end_time - start_time

print(f"Luas Segitiga = {luas_segitiga}")
print(f"Waktu : {waktu:.10f} detik")

Luas Segitiga = 80.0
Waktu : 0.0001637936 detik


### Pemrograman Berbasis Fungsi

In [53]:
import time

def luas_segitiga(alas, tinggi):
    return 0.5 * alas * tinggi

start_time = time.time()

alas = 10
tinggi = 16
luas = luas_segitiga(alas, tinggi)

end_time = time.time()

waktu = end_time - start_time

print(f"Luas Segitiga = {luas}")
print(f"Waktu : {waktu:.10f} detik")

Luas Segitiga = 80.0
Waktu : 0.0001423359 detik


Dengan menggunakan library time maka dapat diketahui perbandingan antara pemrogarman berorientasi objek dan pemrograman fungsional memiliki selisih waktu yang tidak berbeda jauh dan selalu berubah-ubah (tidak tepat) dengan yang tercepat adalah pemrograman berbasis fungsi (0.0001423359 detik) dan pemrograman berorientasi objek (0.0001637936 detik).

## 5. Diberikan Algoritma Metode Bisection berikut:
1. Start
2. Define function f(x)
3. Choose initial guesses x0 and x1 such that f(x0)f(x1) < 0
4. Choose pre-specified tolarabe error e
5. Calculate new approximated root as x2 = (x0 + x1)/2
6. Calculate f(x0)f(x2)\
    a. if f(x0)f(x2) < 0 then x0 = x0 and x1 = x2\
    b. if f(x0)f(x2) > 0 then x0 = x2 and x1 = x1\
    c. if f(x0)f(x2) = 0 then go to (8)
7. if |f(x2)| > e then go to (5) otherwise go to (8)
8. Display x2 as root
9. Stop

Buatlah prosedural programming dari persoalan di atas dan ubahlah dengan functional programming. Lakukan dengan batas toleransi 10^-5. Dengan fungsi yang diberikan sebagai berikut: x - 2^x
dan nilai a dan b adalah di antara 0 dan 1 (nilai antara tidak boleh 0 dan 1).

### Prosedural Programming

In [70]:
def f(x):
    return x - 2 ** x

def bisection_method(f, x0, x1, tol):
    if f(x0) * f(x1) > 0:
      print("Tidak ada akar yang berada di antara x0 dan x1")
      return None

    x2 = (x0 + x1) / 2
    while abs(x2) < tol:
        if  f(x0) * f(x2) < 0:
            x1 = x2
        elif f(x0) * f(x2) > 0:
            x0 = x2
        else:
          return x2
    return x2

x0 = 0.000001
x1 = 0.999999
tol = 10 ** -5
bisection = bisection_method(f, x0, x1, tol)
print(bisection)

Tidak ada akar yang berada di antara x0 dan x1
None


### Functional Programming

In [78]:
bisection = (lambda f, x0, x1, tol:
    (lambda x0, x1:
        (lambda x2:
            x2 if abs(f(x2)) < tol else
            x0 if f(x0) * f(x2) < 0 else
            x1 if f(x0) * f(x2) > 0 else
            x2)(
            (x0 + x1) / 2
        )
    )(x0, x1) if f(x0) * f(x1) < 0 else (
        print("Tidak ada akar yang berada di antara x0 dan x1.") or None
    )
)

f = lambda x: x**3 - x - 2
x0 = 0.000001
x1 = 0.999999
tol = 10 ** -5
bisection = bisection(f, x0, x1, tol)
print(bisection)

Tidak ada akar yang berada di antara x0 dan x1.
None
